lib/aiken/rationals.ak

```diff
@@ -0,0 +1,295 @@
+// Rationals
+
+use aiken/math
+
+// Useful for using Rationals in datums/redeemers which cannot have opaque types
+pub type UncheckedRational {
+  numerator: Int,
+  denominator: Int,
+}
+
+// Opaque type used to ensure the sign of the Rational is managed strictly
+// in the numerator
+pub opaque type Rational {
+  numerator: Int,
+  denominator: Int,
+}
+
+// Converts an UncheckedRational to a Rational
+pub fn check_rational(num: UncheckedRational) -> Rational {
+  let UncheckedRational { numerator, denominator } =
+    num
+  div_int(numerator, denominator)
+}
+
+// Multiplication
+pub fn mul(a: Rational, b: Rational) -> Rational {
+  Rational {
+    numerator: a.numerator * b.numerator,
+    denominator: a.denominator * b.denominator,
+  }
+}
+
+test mul_1() {
+  let a =
+    div_int(2, 3) |> mul(div_int(3, 4))
+  a == Rational { numerator: 6, denominator: 12 }
+}
+
+test mul_2() {
+  let a =
+    div_int(-2, 3) |> mul(div_int(-3, 4))
+  a == Rational { numerator: 6, denominator: 12 }
+}
+
+// Division
+pub fn div(a: Rational, b: Rational) -> Rational {
+  recip(b) |> mul(a)
+}
+
+test div_1() {
+  let a =
+    div_int(2, 3) |> div(div_int(3, 4))
+  a == Rational { numerator: 8, denominator: 9 }
+}
+
+test div_2() {
+  let a =
+    div_int(2, 3) |> div(div_int(-3, 4))
+  a == Rational { numerator: -8, denominator: 9 }
+}
+
+// Create a new Rational
+pub fn div_int(numerator: Int, denominator: Int) -> Rational {
+  if denominator < 0 {
+    Rational { numerator: -numerator, denominator: -denominator }
+  } else {
+    Rational { numerator, denominator }
+  }
+}
+
+test div_int_1() {
+  div_int(2, 3) == Rational { numerator: 2, denominator: 3 }
+}
+
+test div_int_2() {
+  div_int(-2, 3) == Rational { numerator: -2, denominator: 3 }
+}
+
+test div_int_3() {
+  div_int(2, -3) == Rational { numerator: -2, denominator: 3 }
+}
+
+test div_int_4() {
+  div_int(2, 4) == Rational { numerator: 2, denominator: 4 }
+}
+
+test div_int_5() {
+  div_int(-2, -3) == Rational { numerator: 2, denominator: 3 }
+}
+
+test div_int_6() {
+  div_int(-2, -4) == Rational { numerator: 2, denominator: 4 }
+}
+
+// Addition
+pub fn add(a: Rational, b: Rational) -> Rational {
+  Rational {
+    numerator: a.numerator * b.denominator + b.numerator * a.denominator,
+    denominator: a.denominator * b.denominator,
+  }
+}
+
+test add_1() {
+  let a =
+    div_int(2, 3) |> add(div_int(3, 4))
+  a == Rational { numerator: 17, denominator: 12 }
+}
+
+// Subtraction
+pub fn sub(a: Rational, b: Rational) -> Rational {
```

Comment:
Again personal preference but I think `subtract` is worth the extra letters for readability than `sub`.

Reply:
I really go back and forth on things like this. I used subtract in my personal implementation of this as well as other longer function names, but I did find myself getting into some awkward looking code post aiken formatting

Reply:
I think once there are traits/typeclasses/operator overloading we should extend the name since you can be more verbose in the names, without cluttering actual expressions.